File: apps/simplex-badge-service/src/BadgeService/Catalog.hs
{-# LANGUAGE DuplicateRecordFields #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE OverloadedStrings #-}

module BadgeService.Catalog
  ( OfferInvalid (..),
    CatalogRefusal (..),
    offerTotal,
    PricedOffer (..),
    priceOffer,
    catalogCurrency,
    defaultCatalog,
  )
where

import Data.List (find)
import Data.Text (Text)
import Data.Time.Clock (UTCTime)
import Data.Word (Word32, Word64, Word8)
import Simplex.Chat.Badges (BadgeType (..))
import Simplex.Chat.Badges.Service (BadgeOffer (..), BadgePrice (..))
import Simplex.Chat.Badges.Types (BadgeItemStatus (..), BadgeOfferId (..), BadgePriceId (..), OfferDiscount (..))
import Simplex.Chat.PaymentService.Types (CurrencyAmount (..))

data OfferInvalid = OIZeroMonths | OIFreeMonthsExceedTerm | OIDiscountTooLarge | OIAmountUnsellable
  deriving (Eq, Show)

-- | The full price is formed before the division, so nothing is rounded part way.
offerTotal :: CurrencyAmount -> Maybe BadgeOffer -> Either OfferInvalid (Word8, CurrencyAmount, CurrencyAmount)
offerTotal (CurrencyAmount p) = \case
  Nothing -> charge 1 (gross 1)
  Just BadgeOffer {months, discount}
    | months == 0 -> Left OIZeroMonths
    | otherwise -> case discount of
        ODFreeMonths f
          | f >= months -> Left OIFreeMonthsExceedTerm
          | otherwise -> charge months (gross (months - f))
        ODDiscount d
          | d >= 100 -> Left OIDiscountTooLarge
          | otherwise -> charge months (gross months * (100 - fromIntegral d) `div` 100)
  where
    gross :: Word8 -> Word64
    gross m = fromIntegral p * fromIntegral m
    maxAmount :: Word64 -- $1,000,000 in minor units
    maxAmount = 100000000
    charge :: Word8 -> Word64 -> Either OfferInvalid (Word8, CurrencyAmount, CurrencyAmount)
    charge m c
      -- both figures, not just the charge: 100000000 a month over 43 months with 42 free
      -- charges 100000000, but the full price wraps a Word32 and the discount underflows
      | c == 0 || c > maxAmount || gross m > maxAmount = Left OIAmountUnsellable
      | otherwise = Right (m, CurrencyAmount (fromIntegral (gross m)), CurrencyAmount (fromIntegral c))

data PricedOffer = PricedOffer
  { poBadgeType :: BadgeType,
    poMonths :: Word8,
    poPrice :: CurrencyAmount,
    poAmount :: CurrencyAmount,
    poCurrency :: Text
  }
  deriving (Eq, Show)

data CatalogRefusal
  = CRUnknownPrice
  | CRDisabledPrice
  | CRUnknownOffer
  | CRDisabledOffer
  | CROfferNotForPrice
  | CRUnsoldBadgeType BadgeType
  | CRUnpriced OfferInvalid
  deriving (Eq, Show)

priceOffer :: [BadgePrice] -> [BadgeOffer] -> BadgePriceId -> Maybe BadgeOfferId -> Either CatalogRefusal PricedOffer
priceOffer prices offers wantedPriceId wantedOfferId = do
  BadgePrice {badgeType, monthPrice, currency, status = priceStatus} <- known CRUnknownPrice (findPrice wantedPriceId)
  soldBadgeType badgeType
  active CRDisabledPrice priceStatus
  chosenOffer <- resolveOffer wantedPriceId wantedOfferId
  (months, price, amount) <- either (Left . CRUnpriced) Right (offerTotal monthPrice chosenOffer)
  pure PricedOffer {poBadgeType = badgeType, poMonths = months, poPrice = price, poAmount = amount, poCurrency = currency}
  where
    known refusal = maybe (Left refusal) Right
    findPrice p = find (\BadgePrice {priceId} -> priceId == p) prices
    findOffer o = find (\BadgeOffer {offerId} -> offerId == o) offers
    soldBadgeType bt
      | bt == BTSupporter || bt == BTLegend = Right ()
      | otherwise = Left (CRUnsoldBadgeType bt)
    active _ BISActive = Right ()
    active _ BISDeprecated = Right ()
    active refusal BISDisabled = Left refusal
    resolveOffer _ Nothing = Right Nothing
    resolveOffer p (Just o) = do
      chosen@BadgeOffer {priceId = offerPriceId, status = offerStatus} <- known CRUnknownOffer (findOffer o)
      active CRDisabledOffer offerStatus
      case offerPriceId of
        Nothing -> Right (Just chosen)
        Just op
          | op == p -> Right (Just chosen)
          | otherwise -> Left CROfferNotForPrice

catalogCurrency :: Text
catalogCurrency = "usd"

-- | Copied from web/src/catalog.ts, and a test checks they still agree. Seeded
-- insert-only, so repricing means a new price id here plus deprecating the old row.
defaultCatalog :: UTCTime -> ([BadgePrice], [BadgeOffer])
defaultCatalog seededAt = (prices, offers)
  where
    prices =
      [ mkPrice "price_supporter" BTSupporter 700,
        mkPrice "price_legend" BTLegend 7000
      ]
    offers =
      [ mkOffer "offer_3m" "price_legend" 3 (ODFreeMonths 1),
        mkOffer "offer_12m" "price_legend" 12 (ODDiscount 50),
        mkOffer "offer_3m_s" "price_supporter" 3 (ODFreeMonths 1),
        mkOffer "offer_12m_s" "price_supporter" 12 (ODDiscount 50)
      ]
    mkPrice :: Text -> BadgeType -> Word32 -> BadgePrice
    mkPrice pId bType mPrice =
      BadgePrice
        { priceId = BadgePriceId pId,
          badgeType = bType,
          monthPrice = CurrencyAmount mPrice,
          currency = catalogCurrency,
          status = BISActive,
          createdAt = seededAt
        }
    mkOffer :: Text -> Text -> Word8 -> OfferDiscount -> BadgeOffer
    mkOffer oId pId mMonths mDiscount =
      BadgeOffer
        { offerId = BadgeOfferId oId,
          priceId = Just (BadgePriceId pId),
          months = mMonths,
          discount = mDiscount,
          status = BISActive,
          createdAt = seededAt
        }

File: apps/simplex-badge-service/web/src/catalog.ts
export type BadgeType = "supporter" | "legend";

export interface Price {
  priceId: string;
  badgeType: BadgeType;
  monthPrice: number; // minor units
  currency: string;
}

export type Discount =
  | { type: "freeMonths"; freeMonths: number }
  | { type: "discount"; discount: number };

export interface Offer {
  offerId: string;
  priceId: string;
  months: number;
  discount: Discount;
}

export interface Catalog {
  prices: readonly Price[];
  offers: readonly Offer[];
}

export const CATALOG: Catalog = {
  prices: [
    { priceId: "price_supporter", badgeType: "supporter", monthPrice: 700, currency: "usd" },
    { priceId: "price_legend", badgeType: "legend", monthPrice: 7000, currency: "usd" },
  ],
  offers: [
    { offerId: "offer_3m", priceId: "price_legend", months: 3, discount: { type: "freeMonths", freeMonths: 1 } },
    { offerId: "offer_12m", priceId: "price_legend", months: 12, discount: { type: "discount", discount: 50 } },
    { offerId: "offer_3m_s", priceId: "price_supporter", months: 3, discount: { type: "freeMonths", freeMonths: 1 } },
    { offerId: "offer_12m_s", priceId: "price_supporter", months: 12, discount: { type: "discount", discount: 50 } },
  ],
};

// A key in the browser only: it names no row in CATALOG.offers, and the request must be
// sent without an offerId. An empty string cannot do the job, because that is how a
// session says "nothing chosen yet".
export const SINGLE_MONTH = "1m";

export interface Total { months: number; price: number; amount: number }

const MAX_AMOUNT = 100_000_000; // $1,000,000 in minor units

export function offerTotal(monthPrice: number, offer: Offer | undefined): Total | string {
  if (!Number.isInteger(monthPrice) || monthPrice < 0) return "bad month price";
  if (offer === undefined) return charge(1, monthPrice);
  if (!Number.isInteger(offer.months) || offer.months < 0) return "months must be a non-negative integer";
  if (offer.months <= 0) return "zero months";
  const gross = monthPrice * offer.months;
  if (offer.discount.type === "freeMonths") {
    if (!Number.isInteger(offer.discount.freeMonths) || offer.discount.freeMonths < 0) return "free months must be a non-negative integer";
    if (offer.discount.freeMonths >= offer.months) return "free months exceed the term";
    return charge(offer.months, monthPrice * (offer.months - offer.discount.freeMonths), gross);
  }
  if (!Number.isInteger(offer.discount.discount) || offer.discount.discount < 0) return "discount must be a non-negative integer";
  if (offer.discount.discount >= 100) return "discount too large";
  return charge(offer.months, Math.floor((gross * (100 - offer.discount.discount)) / 100), gross);
}

// Both figures, not just the charge: the service refuses a full price over the cap, so a
// total accepted here without that check is sellable on the page and refused at checkout.
function charge(months: number, amount: number, price = amount): Total | string {
  if (amount <= 0 || amount > MAX_AMOUNT || price > MAX_AMOUNT) return "amount unsellable";
  return { months, price, amount };
}

export function savingPercent(price: number, amount: number): number {
  if (price <= 0) return 0;
  return Math.round(((price - amount) / price) * 100);
}
